docs/style-guide.md

```diff
@@ -406,6 +406,91 @@ Examples:
+
+You don't need a `usize` to represent a decimal digit 0-9.  By using the smallest possible integer type, you can reason better about cases where overflow can or cannot occur.
+
+#### Checked Arithmetic
+
+In cases where a hard limit on input is not possible, you can use methods such as [checked_add](https://doc.rust-lang.org/std/primitive.usize.html#method.checked_add).
+
+#### Bounds Testing
+
+By default, the `+` operator in Rust will panic upon overflow in debug mode.  You should employ thorough testing of boundaries to ensure that your arithmetic is never able to overflow.
+
+### Don't accept infinite iterators :: suggested
+
+Consider using `enumerate` to avoid an infinite iterator causing integer overflow by checking that the index does not exceed a bound that you set on the input.
+
+Please do not mind that the following example could be written with `fold` instead of `enumerate`.
+
+```rust
+#[non_exhaustive]
+#[derive(Debug, PartialEq)]
+pub enum Error {
+  Limit,
+}
+
+/// iter: must contain fewer than usize::MAX elements
+pub fn count_zeros<T>(iter: T) -> Result<usize, Error>
+where
+  T: Iterator<Item=u8>
+{
+  let mut result: usize = 0;
+  for (i, v) in iter.enumerate() {
+    if i == std::usize::MAX {
+      return Err(Error::Limit)
+    }
+    if v == 0 {
+      result += 1;
+    }
+  }
+  return Ok(result)
+}
+```
+
+# Imports and Configurations
+
+## Features
+
+### Use no_std :: suggested
+
+Main issues: [#77](https://github.com/unicode-org/icu4x/issues/77), [#151](https://github.com/unicode-org/icu4x/issues/151)
+
+Most ICU4X code will not work in [no_std](https://rust-embedded.github.io/book/intro/no-std.html), since memory allocation is very often required to handle edge cases.  Even our most fundamental type, Locale, requires memory allocation.
+
+However, when designing traits and interfaces, we should make them `no_std`-friendly, such that we can more easily expand in this direction more easily in the future.
+
+## Dependencies
+
+### Avoid heavy dependencies :: suggested
+
+Code size is an important factor for portability.  One of the easiest ways to accidentally bloat your code size is to pull in a heavy dependency.
+
+When possible, write your code in such a way as to reduce dependencies, especially dependencies on heavier libraries.  If you need to add a dependency, consider putting it behind a feature flag.
+
+### Avoid `std::collections::HashMap` :: suggested
+
+The standard library HashMap makes bloated binaries.  In order to reduce code size, consider one of these options:
+
+1. If the keys are known ahead of time, use a `struct` or `enum_map`.
+2. If the set is small, use a vector (with linear-time lookup).
+3. If the set is larger, use a sorted vector (with logarithmic-time lookup).
+
+For option 3, note that sorting algorithms are also bloated.  If possible, you should perform sorting offline (e.g., by requiring that data returned by the data provider is already sorted).
```

Comment:
I think `binary_search` approach is working well for small size sorted operations (see variants in Locale).

Reply:
Thanks for pointing me to:

https://github.com/unicode-org/icu4x/blob/82f30162e8098e2c95aeb14aa1112a2fcb0a4273/components/locale/src/parser/langid.rs#L51-L53

Added!